be/src/exec/workgroup/work_group.cpp

```diff
@@ -2,73 +2,393 @@
+                wg->update_select_factor(0 - negative_total_diff_factor * wg->get_diff_factor() /
+                                                     negative_total_diff_factor);
+            } else if (wg->get_diff_factor() > 0) {
+                wg->update_select_factor(negative_total_diff_factor * wg->get_diff_factor() /
+                                         positive_total_diff_factor);
+            }
+        }
+    } else {
+        // if positive_total_diff_factor + negative_total_diff_factor <= 0
+        // This means that there are not enough resources, but some WorkGs are still over-resourced
+        // So we can reduce the proportion of resources in the WorkGroup that are over-resourced
+        // Then increase the proportion of resources for those WorkGs that are under-resourced
+        for (auto const& wg : _io_wgs) {
+            if (wg->get_diff_factor() < 0) {
+                wg->update_select_factor(positive_total_diff_factor * wg->get_diff_factor() /
+                                         negative_total_diff_factor);
+            } else if (wg->get_diff_factor() > 0) {
+                wg->update_select_factor(0 - positive_total_diff_factor * wg->get_diff_factor() /
+                                                     positive_total_diff_factor);
+            }
+        }
+    }
+
+    schedule_io_task();
+    _cur_schedule_num = _cur_schedule_num_period;
+}
+
+void IoWorkGroupQueue::schedule_io_task() {
+    // In a scheduling period, we calculate in advance
+    for (size_t i = 0; i < _cur_schedule_num_period; i++) {
+        size_t idx = get_next_wg_index();
+        _cur_wait_run_wgs[i] = _io_wgs[idx];
+    }
+    _cur_index = 0;
+}
+
+size_t IoWorkGroupQueue::get_next_wg_index() {
+    int index = -1;
+    double total = 0;
+    for (int i = 0; i < _io_wgs.size(); i++) {
+        _io_wgs[i]->update_cur_select_factor(_io_wgs[i]->get_select_factor());
+        total += _io_wgs[i]->get_select_factor();
+        if (index == -1 || _io_wgs[i]->get_cur_select_factor() > _io_wgs[index]->get_cur_select_factor()) {
+            index = i;
+        }
+    }
+    _io_wgs[index]->update_cur_select_factor(0 - total);
+    return index;
 }
 
-WorkGroupPtr WorkGroupManager::pick_next_wg_for_io() {
-    return _wg_io_queue.pick_next();
+WorkGroupPtr IoWorkGroupQueue::get_next_wg() {
+    int index = _cur_index++;
+    if (index < _cur_schedule_num_period) {
+        return _cur_wait_run_wgs[index];
+    }
+
+    return nullptr;
+}
+
+PriorityThreadPool::Task IoWorkGroupQueue::pick_next_task() {
+    std::unique_lock<std::mutex> lock(_global_io_mutex);
+    while (_ready_wgs.empty()) {
+        _cv.wait(lock);
+    }
+    WorkGroupPtr wg = nullptr;
+    do {
+        adjust_weight_if_need();
+        wg = get_next_wg();
+    } while (wg == nullptr || wg->io_task_queue_size() == 0);
+
+    if (wg->io_task_queue_size() == 1) {
+        _ready_wgs.erase(wg);
+    }
+    _total_task_num--;
+
+    return wg->pick_io_task();
```

Comment:
This looks like old code that needs to be submitted and merged back into the new

Reply:
the function is updated now